src/number_theory/fermat4.lean

```diff
@@ -0,0 +1,335 @@
+  by_cases h0 : 0 < c0,
+  { use [a0, b0, c0], tauto },
+  { use [a0, b0, -c0],
+    apply and.intro (neg_of_minimal hf),
+    apply and.intro hc,
+    apply neg_pos.mpr, apply not_le.mp, intro h2,
+    have h3 : 0 < c0, { exact lt_of_le_of_ne h2 (ne.symm (ne_zero hf.1)) },
+    revert h3, exact h0 }
+end
+
+end fermat_42
+
+lemma int.coprime_of_sqr_sum {m r s : ℤ} (h : m = r ^ 2 + s ^ 2) (h2 : int.gcd r s = 1) :
+  int.gcd m r = 1 :=
+begin
+  by_contradiction hg,
+  obtain ⟨p, ⟨hp, hm, hr⟩⟩ := nat.prime.not_coprime_iff_dvd.mp hg,
+  apply nat.prime.not_dvd_one hp,
+  rw ← h2,
+  apply nat.dvd_gcd hr,
+  apply int.coe_nat_dvd_left.mp,
+  apply (or_self _).mp (int.prime.dvd_mul' hp _),
+  rw [← pow_two, (eq_sub_of_add_eq' (eq.symm h))],
+  apply dvd_sub (int.coe_nat_dvd_left.mpr hm),
+  rw pow_two, exact dvd_mul_of_dvd_left (int.coe_nat_dvd_left.mpr hr) r
+end
+
+lemma int.coprime_of_sqr_sum' {m r s : ℤ} (h : m = r ^ 2 + s ^ 2) (h2 : int.gcd r s = 1) :
+  int.gcd m (r * s) = 1 :=
+begin
+  dunfold int.gcd,
+  rw int.nat_abs_mul,
+  apply nat.coprime.mul_right (int.coprime_of_sqr_sum h h2),
+  rw add_comm at h,
+  rw [int.gcd_comm r s] at h2,
+  exact int.coprime_of_sqr_sum h h2
+end
+
+namespace fermat_42
+
+-- If we have a solution to a ^ 4 + b ^ 4 = c ^ 2, we can construct a smaller one. This
+-- implies there can't be a smallest solution.
+lemma not_minimal {a b c : ℤ}
+  (h : minimal a b c) (ha2 : a % 2 = 1) (hc : 0 < c) : false :=
+begin
+  -- Use the fact that a ^ 2, b ^ 2, c form a pythagorean triple to obtain m and n such that
+  -- a ^ 2 = m ^ 2 - n ^ 2, b ^ 2 = 2 * m * n and c = m ^ 2 + n ^ 2
+  -- first the formula:
+  have ht : pythagorean_triple (a ^ 2) (b ^ 2) c,
+  { calc ((a ^ 2) * (a ^ 2)) + ((b ^ 2) * (b ^ 2))
+        = a ^ 4 + b ^ 4 : by ring
+    ... = c ^ 2 : by rw h.1.2.2
+    ... = c * c : by rw pow_two
+  },
+  -- coprime requirement:
+  have h2 : int.gcd (a ^ 2) (b ^ 2) = 1,
+  { change nat.coprime (int.nat_abs (a ^ 2)) (int.nat_abs (b ^ 2)),
+    rw [pow_two, int.nat_abs_mul],
+    apply nat.coprime.mul;
+    { rw [pow_two, int.nat_abs_mul], apply nat.coprime.mul_right;
+      apply (coprime_of_minimal h) } },
+  -- in order to reduce the possibilities we get from the classification of pythagorean triples
+  -- it helps if we know the parity of a ^ 2 (and the sign of c):
+  have ha22 : a ^ 2 % 2 = 1, { rw [pow_two, int.mul_mod, ha2], norm_num },
+  obtain ⟨m, n, ht1, ht2, ht3, ht4, ht5, ht6⟩ :=
+    pythagorean_triple.coprime_classification' ht h2 ha22 hc,
+  -- Now a, n, m form a pythagorean triple and so we can obtain r and s such that
+  -- a = r ^ 2 - s ^ 2, n = 2 * r * s and m = r ^ 2 + s ^ 2
+  -- formula:
+  have htt : pythagorean_triple a n m,
+  { delta pythagorean_triple,
+    rw [← pow_two, ← pow_two, ← pow_two], exact (add_eq_of_eq_sub ht1) },
+  -- a and n are coprime, because a ^ 2 = m ^ 2 - n ^ 2 and m and n are coprime.
+  have h3 : int.gcd a n = 1,
```

Comment:
I think the proof should look like this: coprime m n -> coprime m^2 n^2 -> coprime m^2-n^2 n^2 = coprime a^2 n^2 -> coprime a n. This is much easier to read than what we have here.

Reply:
@paulvanwamelen What do you think of this suggestion?

Reply:
I liked it! Are my changes not close enough to what Kevin suggested?

Reply:
Sorry, I didn't take a close look, because Kevin already did such an extensive review. So I was just checking whether you dealt with all his comments. So, if this is "done", then I'll resolve the conversation again.
My apologies.